src/Microsoft.Data.SqlClient/netcore/src/Microsoft/Data/SqlClient/SqlDiagnosticListener.NetCoreApp.cs

```diff
@@ -3,13 +3,21 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Diagnostics;
+using System.Reflection;
+using System.Runtime.Loader;
 
 namespace Microsoft.Data.SqlClient
 {
     internal sealed class SqlDiagnosticListener : DiagnosticListener
     {
         public SqlDiagnosticListener(string name) : base(name)
         {
+            AssemblyLoadContext.GetLoadContext(Assembly.GetExecutingAssembly()).Unloading += SqlDiagnosticListener_Unloading;
+        }
+
+        private void SqlDiagnosticListener_Unloading(AssemblyLoadContext obj)
+        {
+            Dispose();
```

Comment:
Just a question here to share some thoughts, it might not be even related to this change but I think it is good to have this in mind when implementing `IDisposable` interface : By [Microsoft Best practices documentation](https://docs.microsoft.com/en-us/dotnet/standard/garbage-collection/implementing-dispose#the-dispose-method) should we consider having a better control over Dispose method? the `bool` value passed to the overridden method will determine about disposing `Managed` and `Unmanaged` resource. I have seen the pattern has been used in some class in our driver, but is not consistent. It definitely needs more thought and testing.

Reply:
These classes have no finalizers and hold no unmanaged resources so the distinction between disposing from the finalizer or dispose methods isn't relevant. If we were talking about one of the SafeHandle types then it would be sensible to use the bool parameter pattern. In this case there is no need for the method to even be called dispose or implement IDisposable because we're doing cleanup directly and not with using blocks or cascaded dispose calls. I followed the convention that was there because the method name Dispose has become synonymous with cleanup/finalize semantics for most .net developers and it's a reasonable name to use.